docs/source/meta_init.rst

```diff
@@ -0,0 +1,239 @@
+.. _meta_init-doc:
+
+.. currentmodule:: torch.nn.utils.meta_init
+
+Large Module Materialization
+============================
+Large Model Materialization API enables constructing ``torch.nn.Module``
+instances on the meta device for inspection purposes. It is meant to be used if
+the size of a model is too big to fit on a single machine and you want to
+inspect it without instantiating.
+
+.. warning::
+    This is an experimental feature and is subject to change. If you experience
+    any issues, let us know by opening a GitHub issue.
+
+
+Problem
+-------
+With ever increasing model sizes, it is becoming increasingly common for models
+to exceed the memory capacity of a single machine. This means training such
+models requires some partitioning strategy to distribute parts of the model onto
+different machines. However the techniques (e.g. 3D parallelism, FSDP) used to
+apply these strategies often need access to the model architecture to decide on
+the optimal strategy and this represents a chicken-egg problem. Partitioning the
+model requires first instantiating the model for inspection purposes and this
+contradicts the whole purpose of partitioning.
+
+Meta Initialization
+-------------------
+The :func:`meta_init` context manager is a non-intrusive API that enables users
+to construct a ``torch.nn.Module`` on the meta device. Internally it forces all
+parameters, buffers, and other tensors within its scope to use the meta device
+regardless of their real device. The returned module(s) can be used for
+inspection purposes and afterwards, if desired, be materialized in a
+fine-grained way via :func:`materialize`.
+
+Note though that the context manager uses a best effort algorithm and is not
+guaranteed to succeed if a module's implementation does not support the meta
+device. If you are the module author, you can use the :func:`is_meta_init()`
+function to find out whether your module is being used within the scope of a
+meta-init context and have an alternate logic if necessary. See
+:ref:`common-failures` below for more information.
+
+Code Examples
+-------------
+The most straightforward use case is to construct a module within the scope of
```

Comment:
Original version of this PR was just recreating the module python object from scratch on the target device as the means of initialization. Why is the pivot towards effectively lazy tensor? Do we have indication that such approach even works on most common models?

Also, how is it going to compose with sharding? It's crucial to think through an e2e example

Reply:
This fine-grained materialization was explicitly requested by PyTorch Lightning, DeepSpeed, and our FSDP project. In fact if you check Lightning, they already use this approach [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/3fcfd0214cc95fa7de49f6404421b5c6cd113c57/pytorch_lightning/trainer/trainer.py#L1424) which is simply a "hacky" workaround built on top of my early draft PR.

I have tested submodule/fine-grained materialization with the majority of the HG models, minGPT (explicitly asked by OpenAI), FAIR's xFormers, and TorchVision. So far haven't observed any issues. Those are models that follow most of the best practices though. Once we release this as an experimental feature, we will most likely see models using some esoteric ways to initialize their states.

TBH I haven't tested this with Pritam's ShardedTensor work so far. We will spend sometime this week to see how they play out together.

Reply:
Adding to the conversation: https://devblog.pytorchlightning.ai/prevent-oom-errors-training-large-models-using-lightning-and-pytorch-meta-tensors-part-ii-83ecfd218186